Save, Read and Delete Audio files,and add Audio Visualizer for display

crazypoo opened the pull request:
Hello friend,#8 complete,and you can see my demo or project https://github.com/crazypoo/PTChatGPT

crazypoo committed:
#8 complete

#8 complete

AppDevGuy commented:
Hi @crazypoo thanks for the contribution. I like that you have tackled something I want to implement. 

Could you please put some details on the Pull Request like:

- What it does.
- Whether you can write some unit tests for the changes it. 
- Do current unit tests still run? (press cmd + shift + U)

I will create a development branch so I can clean it up and contribute a few other things to this. I will also check out your repo. Thanks again.

AppDevGuy commented on OSSSpeechKit/Classes/OSSSpeech.swift:
```diff
@@ -150,6 +150,8 @@ public enum OSSSpeechRecognitionTaskType: Int {
 public protocol OSSSpeechDelegate: AnyObject {
     /// When the microphone has finished accepting audio, this delegate will be called with the final best text output.
     func didFinishListening(withText text: String)
+    ///When the microphone has finished accepting audio, this delegate will be called with the final best text output or voice file path.
```
```suggestion
    /// When the microphone has finished recording audio, this function will be called with the final best text output or voice file path.
```

AppDevGuy commented on OSSSpeechKit/Classes/OSSSpeech.swift:
```diff
@@ -150,6 +150,8 @@ public enum OSSSpeechRecognitionTaskType: Int {
     /// When the microphone has finished accepting audio, this delegate will be called with the final best text output.
     func didFinishListening(withText text: String)
+    ///When the microphone has finished accepting audio, this delegate will be called with the final best text output or voice file path.
+    func didFinishListening(withAudioFileURL url:URL,withText text:String)
```
```suggestion
    func didFinishListening(withAudioFileURL url: URL, withText text: String)
```

AppDevGuy commented on OSSSpeechKit/Classes/OSSSpeech.swift:
```diff
@@ -165,6 +167,9 @@ public class OSSSpeech: NSObject {
 
     // MARK: - Private Properties
 
+    private var audioRecorder:AVAudioRecorder?
```
```suggestion
    private var audioRecorder: AVAudioRecorder?
```

AppDevGuy replied:
Also could you please add a code comment about why we need this?

AppDevGuy commented on OSSSpeechKit/Classes/OSSSpeech.swift:
```diff
@@ -165,6 +167,9 @@ public class OSSSpeech: NSObject {
     // MARK: - Private Properties
 
+    private var audioRecorder:AVAudioRecorder?
+    private var audioFileURL:URL!
```
```suggestion
    private var audioFileURL: URL?
```

AppDevGuy replied:
Could you please add a code comment about why we need this? Also, I have made it an optional since we don't know if it will ever get created or used, so if we attempt to access it, it might cause a crash if it has no value.

AppDevGuy commented on OSSSpeechKit/Classes/OSSSpeech.swift:
```diff
@@ -535,6 +544,40 @@ public class OSSSpeech: NSObject {
+        readyToRecord()
+    }
+    
+    func readyToRecord()
```
Should this be a private function? Or will apps that use this pod be able to access this? 

Could you please add code doc comments (`///`) about what it does?

AppDevGuy commented on OSSSpeechKit/Classes/OSSSpeech.swift:
```diff
@@ -568,6 +612,17 @@ extension OSSSpeech: SFSpeechRecognitionTaskDelegate, SFSpeechRecognizerDelegate
+}
 
+//MARK: AVAudioRecorderDelegate
+extension OSSSpeech:AVAudioRecorderDelegate
```
```suggestion
extension OSSSpeech: AVAudioRecorderDelegate {
```

AppDevGuy commented on OSSSpeechKit/Classes/OSSSpeech.swift:
```diff
@@ -568,6 +612,17 @@ extension OSSSpeech: SFSpeechRecognitionTaskDelegate, SFSpeechRecognizerDelegate
+        if flag
+        {
+            audioRecorder?.stop()
+            print("Audio file save")
```
Should we call the delegate here to return the audioRecording file URL? 

Also, please remove the `print` statement. `print` statements are not usually kept in pull requests,

crazypoo replied:
you want to return the file path only?

AppDevGuy commented on OSSSpeechKit/Classes/OSSSpeech.swift:
```diff
@@ -568,6 +612,17 @@ extension OSSSpeech: SFSpeechRecognitionTaskDelegate, SFSpeechRecognizerDelegate
+extension OSSSpeech:AVAudioRecorderDelegate
+{
+    public func audioRecorderDidFinishRecording(_ recorder: AVAudioRecorder, successfully flag: Bool) {
+        if flag
```
```suggestion
        if flag {
```

AppDevGuy commented on OSSSpeechKit/Classes/OSSSpeech.swift:
```diff
@@ -568,6 +612,17 @@ extension OSSSpeech: SFSpeechRecognitionTaskDelegate, SFSpeechRecognizerDelegate
     public func speechRecognizer(_ speechRecognizer: SFSpeechRecognizer, availabilityDidChange available: Bool) {}
+}
 
+//MARK: AVAudioRecorderDelegate
```
```suggestion
// MARK: AVAudioRecorderDelegate

```

AppDevGuy commented on OSSSpeechKit/Classes/OSSSpeech.swift:
```diff
@@ -535,6 +544,40 @@ public class OSSSpeech: NSObject {
+            audioRecorder = try AVAudioRecorder(url: audioFileURL, settings: audioSettings)
+            audioRecorder?.delegate = self
+            audioRecorder?.record()
+
```
Please remove the blank line from here.

AppDevGuy commented on OSSSpeechKit/Classes/OSSSpeech.swift:
```diff
@@ -535,6 +544,40 @@ public class OSSSpeech: NSObject {
+            audioRecorder?.record()
+
+        }
+        catch
```
```suggestion
        } catch {
```

AppDevGuy commented on OSSSpeechKit/Classes/OSSSpeech.swift:
```diff
@@ -535,6 +544,40 @@ public class OSSSpeech: NSObject {
+        }
+        catch
+        {
+            print(error.localizedDescription)
```
We should add a new `OSSSpeechKitErrorType` and then use the `didFailToProcessRequest(withError` function to handle sharing the error.

AppDevGuy commented on OSSSpeechKit/Classes/OSSSpeech.swift:
```diff
@@ -535,6 +544,40 @@ public class OSSSpeech: NSObject {
+        dateFormatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
+        let dateString = dateFormatter.string(from: Date())
+        
+        self.audioFileURL = getDocumentsDirectory().appendingPathComponent("\(dateString).m4a")
```
```suggestion
guard audioFileURL = getDocumentsDirectory().appendingPathComponent("\(dateString).m4a") else {
    // use the delegate to handle the error and return.
}
```

AppDevGuy reviewed (changes requested):
Thanks again for submitting a pull request. I hope my comments help you improve your work, it's a good start. 

I would also like it if instead of recording every time, we use a `public` variable that allows the developer or user to decide if they want to record every time. We don't want to record and save to the documents directory every single time, as that could result in making their app very big. 

I also would like to see a function that deletes recordings too because if we are responsible for saving the recordings, we must also be responsible for enabling them to be deleted. 

I hope this helps. Thanks again.

crazypoo commented:
ok,i will fix it

crazypoo committed:
修復一些昆蟲

修復一些昆蟲

crazypoo committed:
Can delete voice file

1.主要添加删除用户声音的方法/主要添加刪除用戶聲音的方法/Add methods to delete user voice file(s)/Añadir métodos para eliminar el(los) archivo(s) de voz del usuario
2.添加了可以选择候录音的选项/添加了可以選擇錄音的選項/Added the option to wait for recording/Se ha añadido la opción de esperar la grabación
3.修复了一些昆虫/修復了一些昆蟲/Fix some bugs/Corregir algunos errores

crazypoo committed:
Transcription voice file path

Transcription voice file path

crazypoo committed:
bug fix

crazypoo committed:
Add display sound ripple parameter callback for more UI display

crazypoo committed:
bug fix

hllGitHub commented:
Can this pull request be accepted? I hope I can use this feature, thanks.

crazypoo commented:
> Can this pull request be accepted? I hope I can use this feature, thanks.

我放在我这里了,以后在我这里更新,https://github.com/crazypoo/PTools

huangboju commented:
@crazypoo 
Actually, there's no need to create an AVAudioRecorder instance, since AVAudioEngine can already capture audio volume levels.
For example：

```
input.installTap(onBus: bus, bufferSize: 8192, format: recordingFormat) { [weak self] (buffer, _) -> Void in
            var newBufferAvailable = true
            let inputCallback: AVAudioConverterInputBlock = { _, outStatus in
                if newBufferAvailable {
                    outStatus.pointee = .haveData
                    newBufferAvailable = false
                    return buffer
                } else {
                    outStatus.pointee = .noDataNow
                    return nil
                }
            }
            let convertedBuffer = AVAudioPCMBuffer(pcmFormat: outputFormat, frameCapacity: AVAudioFrameCount(outputFormat.sampleRate) * buffer.frameLength / AVAudioFrameCount(buffer.format.sampleRate))!
            var error: NSError?
            let status = converter.convert(to: convertedBuffer, error: &error, withInputFrom: inputCallback)
            if status == .error {
                self?.delegate?.didFailToCommenceSpeechRecording()
                self?.delegate?.didFailToProcessRequest(withError: OSSSpeechKitErrorType.invalidAudioEngine.error)
                if let err = error {
                    self?.debugLog(object: self as Any, message: "Audio Engine conversion error: \(err)")
                }
                return
            }
            self?.request?.append(convertedBuffer)
            self?.processAudioBuffer(convertedBuffer)
        }

    private func processAudioBuffer(_ buffer: AVAudioPCMBuffer) {
        guard let channelData = buffer.floatChannelData?[0] else { return }

        let channelDataValue = channelData
        let channelDataValueArray = stride(from: 0, to: Int(buffer.frameLength), by: buffer.stride).map { channelDataValue[$0] }

        // 计算RMS (Root Mean Square) 音量
        let rms = sqrt(channelDataValueArray.map { $0 * $0 }.reduce(0, +) / Float(channelDataValueArray.count))

        // 转换为分贝并标准化
        let decibels = min(max(20 * log10(rms), -60), 0)

        let normalizedVolume = (decibels + 60) / 60.0 // 标准化到 0-1 范围

        DispatchQueue.main.async {
            self.delegate?.didProcessAudioBuffer(normalizedVolume)
        }
    }
```
